balancing/balance_breaker.go

```diff
@@ -0,0 +1,615 @@
+}
+
+// SetActive sets meter state
+func (meter *CallMeter) SetActive(active bool) {
+	if meter.IsActive() && !active {
+		meter.inActiveSince = meter.now()
+	}
+	if !meter.IsActive() && active {
+		meter.histogram.shiftData(meter.now().Sub(meter.inActiveSince))
+		meter.inActiveSince = time.Time{}
+	}
+}
+
+// TimeSpent returns float64 repesentation of time spent in execution
+func (meter *CallMeter) TimeSpent() float64 {
+	allSeries := meter.histogram.pickLastSeries(meter.resolution)
+	sum := float64(0)
+	now := meter.now()
+
+	for _, series := range allSeries {
+		series.ValueRangeFun(now.Add(-meter.resolution), now, func(value *timeValue) {
+			sum += value.value
+		})
+	}
+
+	return sum
+}
+
+type dataSeries struct {
+	data []*timeValue
+	mx   sync.Mutex
+}
+
+func (series *dataSeries) Add(value float64, dateTime time.Time) {
+	series.mx.Lock()
+	defer series.mx.Unlock()
+	series.data = append(series.data, &timeValue{dateTime, value})
+}
+
+func (series *dataSeries) ValueRangeFun(timeStart, timeEnd time.Time, fun func(*timeValue)) []float64 {
+	dataRange := []float64{}
+	for _, timeVal := range series.data {
+		if (timeStart == timeVal.date || timeStart.Before(timeVal.date)) && timeEnd.After(timeVal.date) {
+			fun(timeVal)
+		}
+	}
+	return dataRange
+}
+
+func (series *dataSeries) ValueRange(timeStart, timeEnd time.Time) []float64 {
+	dataRange := []float64{}
+	series.ValueRangeFun(timeStart, timeEnd, func(value *timeValue) {
+		dataRange = append(dataRange, value.value)
+	})
+	return dataRange
+}
+
+type timeValue struct {
+	date  time.Time
+	value float64
+}
+
+func newTimeHistogram(retention, resolution time.Duration, now func() time.Time) *histogram {
+	return &histogram{
+		t0:         now(),
+		resolution: resolution,
+		retention:  retention,
+		now:        now,
+		mx:         sync.Mutex{},
+	}
+}
+
+type histogram struct {
+	t0         time.Time
+	retention  time.Duration
+	resolution time.Duration
+	data       []*dataSeries
+	now        func() time.Time
+	mx         sync.Mutex
+}
+
+func (h *histogram) pickSeries(at time.Time) *dataSeries {
+	h.mx.Lock()
+	defer h.mx.Unlock()
+	idx := h.index(at)
+	if idx < 0 {
+		return nil
+	}
+	if idx >= h.cellsCount() || idx >= len(h.data) {
+		h.unshiftData(at)
+		idx = h.index(at)
+	}
+	return h.data[idx]
+}
+
+func (h *histogram) pickLastSeries(period time.Duration) []*dataSeries {
+	h.mx.Lock()
+	defer h.mx.Unlock()
+	if period > h.retention {
+		period = h.retention
+	}
+	h.unshiftData(h.now())
+	cellsNumber := math.Ceil(float64(period)/float64(h.resolution)) + 1
+	now := h.now()
+	stop := h.index(now) + 1
+	start := int(math.Max(float64(stop-int(cellsNumber)), 0))
+	return h.data[start:stop]
+}
+
+func (h *histogram) index(now time.Time) int {
+	sinceStart := float64(now.Sub(h.t0))
+	idx := math.Floor(sinceStart / float64(h.resolution))
+	return int(idx)
+}
+
+func (h *histogram) cellsCount() int {
+	return int(math.Ceil(float64(h.retention)/float64(h.resolution))) + 1
+}
+
```

Comment:
Some of the methods of `histogram` are protected by a mutex, but some are not. `shiftData` mutates the state heavily, yet it doesn't lock the data it mutates. If (or when)`shiftData` and `unshiftData` run concurrently bad things may happen.